pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,244 @@
+			break
+		}
+
+		if i == retries {
+			break // it was the last attempt
+		}
+
+		backoff := time.Duration(multiplier) * time.Second
+		time.Sleep(backoff)
+		multiplier *= 2
+	}
+
+	return conn, err
+}
+
+type downwardMetricsServer struct {
+	retries            uint // Number of retires after the first connection fails
+	virtioSerialSocket string
+	scraper            *metricsScraper.Scraper
+	libvirtSocket      string
+}
+
+func (s *downwardMetricsServer) handleRequest(conn net.Conn, requestLine string) error {
+	requestPath, err := parseRequestLine(requestLine)
+	if err != nil {
+		replyError(conn)
+		return err
+	}
+
+	if requestPath != "/metrics/XML" {
+		replyError(conn)
+		return errors.New("invalid request")
+	}
+
+	var xmlMetrics []byte
+	metrics, err := s.scraper.Report(s.libvirtSocket)
+	if err != nil {
+		xmlMetrics = emptyMetrics
+		log.Log.Reason(err).Warningf("failed to collect the metrics")
+	} else {
+		xmlMetrics, err = xml.MarshalIndent(metrics, "", "  ")
+		if err != nil {
+			return fmt.Errorf("failed to encode metrics: %v", err)
+		}
+	}
+
+	// `vm-dump-metrics` expects `\n\n` as a termination symbol
+	xmlMetrics = append(xmlMetrics, '\n', '\n')
+
+	err = reply(conn, xmlMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to send the metrics: %v", err)
+	}
+	return nil
+}
+
+// DownwardMetrics Virtio Server
+// By defining a virtio-serial device, QEMU creates a unix domain socket on
+// the host that is connected to a serial port in the VM.
+//
+// This server is a bit special, since it serves the process, inside the VM,
+// requests by replying with the DownwardMetrics. But it connects to the
+// socket at the client end, since the unix socket is created by QEMU.
+func (s *downwardMetricsServer) start() error {
+	err := waitForSocket(s.virtioSerialSocket)
+	if err != nil {
+		return err
+	}
+
+	conn, err := connect(s.virtioSerialSocket, s.retries)
+	if err != nil {
+		return err
+	}
+	defer func(conn net.Conn) {
+		_ = conn.Close()
+	}(conn)
+
+	for {
+		request, err := waitForRequest(conn)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			return err
+		}
+
+		go func(conn net.Conn, request string) {
+			err = s.handleRequest(conn, request)
+			if err != nil {
+				log.Log.Reason(err).Errorf("failed to process the request")
+			}
+		}(conn, request)
+	}
+}
+
+func RunDownwardMetricsVirtioServer(nodeName string) error {
+	exists, err := diskutils.FileExists(standardLauncherSocketPath)
+	if !exists || err != nil {
+		log.Log.Reason(err).Errorf("virt-launcher socket not found")
+		return err
+	}
+	isolationDetector := isolation.NewSocketBasedIsolationDetector(util.VirtShareDir)
+	scraper := metricsScraper.NewScraper(nodeName, isolationDetector)
+
+	server := downwardMetricsServer{
+		retries:            5,
+		virtioSerialSocket: downwardmetrics.DownwardMetricsChannelSocket,
+		scraper:            scraper,
+		libvirtSocket:      standardLauncherSocketPath,
+	}
+
+	go func() {
```

Comment:
There is no way to cleanly shutdown the server like this? You should at least add a done channel.

Reply:
How do you signal to the outside that the server possibly failed (e.g. encountered EOF above)?

Reply:
Fixed